pkg/vault/contracts/Vault.sol

```diff
@@ -1336,23 +1344,24 @@ contract Vault is IVaultMain, VaultCommon, Proxy {
                 // EXACT_IN requires the exact amount of wrapped tokens to be unwrapped, so we call redeem.
                 // The amount of wrapped tokens to redeem is the amount necessary to fulfill the trade
                 // (amountInWrapped), plus the amount needed to leave the buffer rebalanced 50/50 at the end
-                // (bufferWrappedSurplus).
-                vaultWrappedDeltaHint = amountInWrapped + bufferWrappedSurplus;
+                // (bufferWrappedToRebalance). `bufferWrappedToRebalance` may be positive if buffer has an excess of
+                // wrapped, or negative if the buffer has an excess of underlying tokens. `vaultWrappedDeltaHint`
+                // will always be a positive number, because if `abs(bufferWrappedToRebalance) > amountInWrapped` and
+                // `bufferWrappedToRebalance < 0`, it means the buffer has enough liquidity to fulfill the trade (i.e.
+                // `bufferBalances.getBalanceRaw() >= amountOutUnderlying` is true).
+                vaultWrappedDeltaHint = (amountInWrapped.toInt256() + bufferWrappedToRebalance).toUint256();
                 vaultUnderlyingDeltaHint = wrappedToken.redeem(vaultWrappedDeltaHint, address(this), address(this));
             } else {
                 // EXACT_OUT requires the exact amount of underlying tokens to be returned, so we call withdraw.
-                // The amount of underlying tokens to withdraw is the amount necessary to fulfill the trade
-                // (amountOutUnderlying), plus the amount needed to leave the buffer rebalanced 50/50 at the end
-                // (bufferUnderlyingSurplus).
-                if (bufferWrappedSurplus > 0) {
-                    // An unwrap operation withdraws wrapped tokens from the buffer and returns underlying tokens.
-                    // For the buffer to rebalance during an unwrap operation, it must have an excess (surplus) of
-                    // wrapped tokens. Since this is an EXACT_OUT unwrap (withdraw), we need to calculate the amount of
-                    // underlying tokens the buffer should receive to rebalance, based on the wrapped token surplus.
-                    // The redeem function is used here, as it performs the inverse of a withdraw operation.
-                    bufferUnderlyingSurplus = wrappedToken.previewRedeem(bufferWrappedSurplus);
+                // The amount of underlying tokens to withdraw is calculated in terms of the wra.
+                if (bufferWrappedToRebalance != 0) {
```

Comment:
For symmetry, should have a comment inside the if corresponding to `The amount of underlying tokens to deposit is the necessary amount to fulfill the trade...` in `_wrapWithBuffer`.